src/libcore/array.rs

```diff
@@ -285,3 +304,171 @@ macro_rules! array_impl_default {
 }
 
 array_impl_default!{32, T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T}
+
+
+/// An iterator that moves out of an array.
+///
+/// This `struct` is created by the `into_iter` method on [arrays]
+/// (provided by the [`IntoIterator`] trait).
+///
+/// [arrays]: ../../std/primitive.array.html
+/// [`IntoIterator`]: ../../std/iter/trait.IntoIterator.html
+#[unstable(feature = "array_into_iter", issue = "0")]
+pub struct IntoIter<T, A: FixedSizeArray<T>> {
+    // Invariants: index <= index_back <= array.len()
+    // Only values in array[index..index_back] are alive at any given time.
+    // Values from array[..index] and array[index_back..] are already moved/dropped.
+    array: ManuallyDrop<A>,
+    index: usize,
+    index_back: usize,
+    _marker: PhantomData<T>,
+}
+
+impl<T, A: FixedSizeArray<T>> IntoIter<T, A> {
+    /// Returns the remaining items of this iterator as a slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(array_into_iter)]
+    ///
+    /// # fn main() {
+    /// let array = ['a', 'b', 'c'];
+    /// let mut into_iter = array.into_iter();
+    /// assert_eq!(into_iter.as_slice(), &['a', 'b', 'c']);
+    /// let _ = into_iter.next().unwrap();
+    /// assert_eq!(into_iter.as_slice(), &['b', 'c']);
+    /// # }
+    /// ```
+    #[inline]
+    #[unstable(feature = "array_into_iter", issue = "0")]
+    pub fn as_slice(&self) -> &[T] {
+        &self.array.as_slice()[self.index..self.index_back]
+    }
+
+    /// Returns the remaining items of this iterator as a mutable slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(array_into_iter)]
+    ///
+    /// # fn main() {
+    /// let array = ['a', 'b', 'c'];
+    /// let mut into_iter = array.into_iter();
+    /// assert_eq!(into_iter.as_slice(), &['a', 'b', 'c']);
+    /// into_iter.as_mut_slice()[2] = 'z';
+    /// assert_eq!(into_iter.next().unwrap(), 'a');
+    /// assert_eq!(into_iter.next().unwrap(), 'b');
+    /// assert_eq!(into_iter.next().unwrap(), 'z');
+    /// # }
+    /// ```
+    #[inline]
+    #[unstable(feature = "array_into_iter", issue = "0")]
+    pub fn as_mut_slice(&mut self) -> &mut [T] {
+        &mut self.array.as_mut_slice()[self.index..self.index_back]
```

Comment:
Same question here-- I think this implementation should use `ptr::offset` and `from_raw_parts` so as not to acquire a reference to a partially uninitialized slice.

Reply:
Same question as above, how can I get a base pointer at all without forming some reference?

Reply:
As above-- if you make `ManuallyDrop` `repr(transparent)`, then you know that pointers to it will point directly to its element.